File: source/python/barrier_sync.py
import threading

num_tasks = 4
barrier = threading.Barrier(num_tasks)
done_count = 0
lock = threading.Lock()

def barrier_task(task_id):
    global done_count
    barrier.wait()
    with lock:
        done_count += 1

threads = []
for i in range(num_tasks):
    t = threading.Thread(target=barrier_task, args=(i,))
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print(f"all {done_count} tasks passed barrier")

File: source/rust/barrier_sync.rs
use std::sync::{Arc, Barrier};
use std::thread;

fn main() {
    let num_tasks = 4;
    let barrier = Arc::new(Barrier::new(num_tasks));
    let (tx, rx) = std::sync::mpsc::channel();

    let mut handles = Vec::new();
    for i in 0..num_tasks {
        let b = Arc::clone(&barrier);
        let tx = tx.clone();
        handles.push(thread::spawn(move || {
            b.wait();
            tx.send(i).unwrap();
        }));
    }
    drop(tx);

    for h in handles { h.join().unwrap(); }
    let count = rx.into_iter().count();
    println!("all {} tasks passed barrier", count);
}
